Add support for writing elemental variables in Nemesis

jwpeterson opened the pull request:
Also add test of the new capability, and update (re-gold) existing Nemesis tests that had elemental variables but were not writing them previously. This change requires a libMesh update, which is briefly summarized below.

* Add support for elemental variable writing in Nemesis.
* Add support for different node partitioning strategies (@fdkong)
* configure: detect when PETSc is built with support for different partitioners
* Drop usage of `libmesh_nullptr`, `libmesh_final`, `libmesh_override` etc. from library.
* Allow partitioners to refer to the "original" based on the new local ordering.
* Default/delete copy/move constructors and assignment operators in various classes; add unit tests.
* Add PostWaitDereferenceTag class and reference counting for MessageTags.
* Replace old "round-robin" communication algorithm with parallel_sync in several places.
* Use MessageTag in parallel_sync probe calls.
* NumericVector::operator/= argument is now const reference.
* configure: Improve support for PGI compiler (@RonRahaman)
* configure: Add support for --enable-march and --with-gcc-arch.
* Fix ptrdiff_t/ std::ptrdiff_t incompatibility issue.
* Fix overzealous assertion in UnstructuredMesh::copy_nodes_and_elements() by wrapping proc ids.
* Switch all enumerations to "fixed type" enums with underlying type int.
* configure: disable HDF5 detection by default.
* Fix issue with --use-petsc-dm (@fdkong).
* Add Elem::nodes_on_side() utility function (@lindsayad)
* Fix/ignore misc. new warnings.
* Add --require-petsc-hypre configure option.
* C++11 decltype support is now required.

Refs libMesh/libmesh#1764.

jwpeterson committed:
Add test of elemental variable writing in Nemesis.

We have two version of this test: one for ReplicatedMesh runs and one
for DistributedMesh runs. The resulting gold files have different node
numbering and therefore different partitioning, so they can't be shared
by both tests.

Refs libMesh/libmesh#1764

jwpeterson committed:
libMesh update.

Brief summary of changes in this update:
* Add support for elemental variable writing in Nemeis.
* Add support for different node partitioning strategies (@fdkong)
* configure: detect when PETSc is built with support for different partitioners
* Drop usage of libmesh_nullptr, libmesh_final, libmesh_override etc. from library.
* Allow partitioners to refer to the "original" based on the new local ordering.
* Default/delete copy/move constructors and assignment operators in various classes; add unit tests.
* Add PostWaitDereferenceTag class and reference counting for MessageTags.
* Replace old "round-robin" communication algorithm with parallel_sync.
* Use MessageTag in parallel_sync probe calls.
* NumericVector::operator/= argument is now const reference.
* configure: Improve support for PGI compiler (@RonRahaman)
* configure: Add support for --enable-march and --with-gcc-arch.
* Fix ptrdiff_t/ std::ptrdiff_t incompatibility issue.
* Fix overzealous assertion in UnstructuredMesh::copy_nodes_and_elements() by wrapping proc ids.
* Switch all enumerations to "fixed type" enums with underlying type int.
* configure: disable HDF5 detection by default.
* Fix issue with --use-petsc-dm (@fdkong).

Refs #000.

jwpeterson committed:
Nemeis: Add call to write_element_data().

This requires a libmesh update.

jwpeterson committed:
Update/regold the nemesis_hide tests.

These tests have a CONSTANT, MONOMIAL AuxVariables which are now being
(correctly) written out, so the gold file needs to be updated. The
AuxVariable previously had no AuxKernel assigned to it, so that has
now been fixed as well.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
+Subproject commit 000bf49e1dbc54444cf1ca3d06791453922d69e0
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

jwpeterson committed:
Fixes for non-unity builds.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/204966/) on 77fb027 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11754/site)

This comment will be updated on new commits.

jwpeterson commented:
Looks like we have picked up some DistributedMesh regressions in this libMesh update. They all look to be timeouts rather than crashing, and several seem to involve adaptivity. 

The code could be hanging in one of the new parallel_sync calls introduced by libMesh/libmesh#1708, or related to the request message tags stuff introduced in libMesh/libmesh#1741 and libMesh/libmesh#1748. I will try to get some backtraces and/or bisect in libmesh to help figure out what's going on...

```
    FAIL executioners/executioner.test_steady_adapt FAILED (TIMEOUT)
    FAIL misc/serialized_solution.adapt FAILED (TIMEOUT)
    FAIL time_integrators/actually_explicit_euler_verification.2d-linear-adapt FAILED (TIMEOUT)
    FAIL time_integrators/bdf2.adapt FAILED (TIMEOUT)
    FAIL userobjects/toggle_mesh_adaptivity.toggle_mesh_adaptivity_wait FAILED (TIMEOUT)
    FAIL adaptivity/cycles_per_step.test FAILED (TIMEOUT) [FINISHED]
    FAIL outputs/oversample.ex02_adapt FAILED (TIMEOUT)
    FAIL outputs/oversample.adapt FAILED (TIMEOUT)
    FAIL markers/error_tolerance_marker.adapt_test FAILED (TIMEOUT)
    FAIL adaptivity/max_h_level.test FAILED (TIMEOUT)
    FAIL userobjects/geometry_snap.geometrysphere FAILED (TIMEOUT)
    FAIL markers/two_circle_marker.two_circle_marker_gaussian_ic FAILED (TIMEOUT) [FINISHED]
    FAIL controls/time_periods/dgkernels.test FAILED (TIMEOUT)
    FAIL dirackernels/point_caching.point_caching_adaptive_refinement FAILED (TIMEOUT)
    FAIL executioners/adapt_and_modify.test FAILED (TIMEOUT)
    FAIL mesh/adapt.test_time FAILED (TIMEOUT)
    FAIL mesh/adapt.interval FAILED (TIMEOUT)
    FAIL mesh/adapt.test FAILED (TIMEOUT) [FINISHED]
    FAIL time_integrators/crank-nicolson.adapt FAILED (TIMEOUT) [FINISHED]
    FAIL preconditioners/pbp.pbp_adapt_test FAILED (TIMEOUT)
    FAIL markers/two_circle_marker.two_circle_marker_coarsen FAILED (TIMEOUT) [FINISHED]
    FAIL postprocessors/mms_slope.test FAILED (TIMEOUT) [FINISHED]
    FAIL markers/error_fraction_marker.no_clear FAILED (TIMEOUT)
    FAIL time_integrators/implicit-euler.adapt FAILED (TIMEOUT) [FINISHED]
    FAIL adaptivity/recompute_markers_during_cycles.test FAILED (TIMEOUT) [FINISHED]
    FAIL dirackernels/point_caching.point_caching_uniform_refinement FAILED (TIMEOUT)
    FAIL adaptivity/initial_marker.test FAILED (TIMEOUT)
    FAIL markers/oriented_box_marker.obm FAILED (TIMEOUT)
    FAIL time_integrators/explicit-euler.2d-linear-adapt FAILED (TIMEOUT)
    FAIL markers/box_marker.mark_and_adapt FAILED (TIMEOUT) [FINISHED]
    FAIL geomsearch/nearest_node_locator.adapt FAILED (TIMEOUT) [FINISHED]
    FAIL time_integrators/dirk.dirk-2d-heat-adap FAILED (TIMEOUT)
```

We'll need to get these ironed out before merging this update can be merged...

jwpeterson committed:
Fix incorrect use of LIBMESH_HAVE_VTK ifdef.

This was broken in 5b39a40 but we never noticed it because Unity
builds covered up the issue. You can't use the LIBMESH_HAVE_VTK
preprocessor macro unless you have included the header that
(potentially) defines it first!

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
+Subproject commit 17cb8046ea2dcf6cf6bc3a250c069cfe925c3bfe
```
**Caution!** This contains a submodule update

jwpeterson commented:
I will clean up the history here if libMesh/libmesh#1767 fixes the hangs we were seeing on parallel, adaptive runs.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
+Subproject commit 92686be03fbb1798cd690ab4a36969c267b8d294
```
**Caution!** This contains a submodule update

jwpeterson committed:
libMesh update.

Last minute update on the heels of de14c1b. Changes in this update
include:

* Add Elem::nodes_on_side() utility function (@lindsayad)
* Fix/ignore misc. new warnings.
* Add --require-petsc-hypre configure option.
* C++11 decltype support is now required.

Refs #000.

jwpeterson committed:
Require both PETSc and Hypre when configuring libmesh.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
+Subproject commit d46058ca7e593e17983588958174f5f6f6d1d864
```
**Caution!** This contains a submodule update
